contracts/eosio.system/src/eosio.system.cpp

```diff
@@ -397,6 +397,94 @@ namespace eosiosystem {
       set_resource_limits( account, current_ram, current_net, cpu );
    }
 
+   // --------------------------------------------------------------------------
+   struct blacklist_manager {
+      blacklist_manager(std::vector<name>& blacklist) : _blacklist(blacklist) {}
+
+      void add(const std::vector<name>& patterns) const {
+         std::unordered_set<name, _hash> already;
+
+         for (auto n : _blacklist)
+            already.insert(n);
+
+         for (auto n : patterns) {
+            if (!already.contains(n))
+               _blacklist.push_back(n);
+         }
+      }
+
+      void remove(const std::vector<name>& patterns) const {
+         std::unordered_set<name, _hash> already;
+
+         for (auto n : _blacklist)
+            already.insert(n);
+
+         for (auto n : patterns) {
+            if (already.contains(n))
+               std::erase(_blacklist, n);
+         }
+      }
+
+   private:
+      // hash function removing least significant zero characters
+      struct _hash {
+         std::size_t operator()(const name& n) const noexcept {
+            uint64_t v = n.value;
+            auto num_zeros = std::countr_zero(v);
+            check(num_zeros >= 4, "bad account name");
+            uint64_t shift = 4 + ((num_zeros - 4) / 5) * 5;
+            return v >> shift;
+         }
+      };
+
+      std::vector<name>& _blacklist;
+   };
+
+   void system_contract::addblnames( const std::vector<name>& blacklisted_name_patterns ) {
+      require_auth( get_self() );
+
+      check(!blacklisted_name_patterns.empty(), "Empty list of blacklisted name patterns provided");
+      check(blacklisted_name_patterns.size() <= 512, "Cannot provide more than 512 patterns in one action call");
```

Comment:
Do we want to enforce a total max size? Seems some limit would be good.

Reply:
Because this is only allowed by a privileged action (multisig), I don't think we should worry about having a total limit.